### ISSUE TITLE
Add seeking functionality to the decoder

### ISSUE THREAD

**Issue by AnthumChris** [AnthumChris/opus-stream-decoder#1]
![seeker](https://user-images.githubusercontent.com/10064176/47256393-b475f600-d44d-11e8-9004-35aeea9edd01.jpg)

Investigate adding a feature like `OpusStreamDecoder.seekTo(milliseconds)` to start decoding at a specific time within the file.  Currently, the decoder only supports decoding a file from the beginning.

Using [`libopusfile.op_pcm_seek()`](https://mf4.xiph.org/jenkins/view/opus/job/opusfile-unix/ws/doc/html/group__stream__seeking.html) seems like a good idea.  `OpusStreamDecoder`'s underlying C implementation may need to dynamically allocate `OpusChunkDecoder.buffer` to hold the entire size of the file to decoded file.  Then, bytes would be enqueued into `buffer` at certain positions for `op_pcm_seek()` to succeed.

Because of Opus' dynamic streaming nature, the total samples/duration of the file would either need to be known beforehand or calculated on the fly (see "[how do I get duration...](https://wiki.xiph.org/OpusFAQ#How_do_I_get_the_duration_of_a_.opus_file.3F)" and "[why don't you store the duration..](https://wiki.xiph.org/OpusFAQ#Why_don.27t_you_store_the_duration_in_the_header.3F_Isn.27t_all_of_that_slow_and_complicated.3F)").

Guesswork would be required to decide which bytes would be required to start playing specific time.  For example, take a 4mb (4,194,304 bytes) file with a duration of 4:00m (240,000 seconds). If we wanted to start playing at the 1:00m mark, which is 25% of the total duration, we could assume that the byte offset to start reading would be near 25% past the start of the file at 1,048,576 bytes (0.25 * 4194304).  We would probably need to pad the beginning of that offset to ensure that the decoder has enough bytes to read packets for the [pre-roll decoding](https://mf4.xiph.org/jenkins/view/opus/job/opusfile-unix/ws/doc/html/group__stream__seeking.html), and assuming an Ogg Opus file's maximum page is 64k, that could potentially be enough.  These are all initial guesses.

**Comment by zandaqo** [comment]
@AnthumChris Thanks, this is exactly what I was looking for. Under this scenario (guessing required bytes), aside from total size and duration, do we need any to have any other data to decode/play the required chunk? For example, do we need the beginning of a file or some other metadata? 

I have a use case where I want to play paragraphs/pages of audiobooks. Currently I have to create a separate file for each paragraph, but I'd rather have one file for a book and load it partially. I'm ok with padding overhead and storing extra metadata.

**Comment by AnthumChris** [comment]
Yes, `OpusStreamDecoder`'s underlying `libopusfile` C library seems to require the beginning of the file, specifically the Ogg header pages and first audio data page. Data page sizes vary based on encoding bit rate (higher bitrate, larger page size). Searching my test file with `xxd` for "OggS" (`0x4f676753`) page boundaries, you can see how a `libopusfile` requires a minimum 8,918 bytes (end of first data page). Anything less fails:
```bash
# prep the test
make test-wasm-module 

# Inspect file.  End of first data page is 0x00022d6 — 8,918 bytes
xxd -l 8950 -g 8 -c 14 tmp/decode-test-64kbps.opus | grep "OggS"

# create truncated file with 8,918 bytes and decode it - succeeds
LEN=8918
dd if=tmp/decode-test-64kbps.opus of=tmp/truncated-$LEN.opus bs=1 count=$LEN
node dist/test-opus-stream-decoder.js tmp/truncated-$LEN.opus tmp

# create truncated file with 8,917 bytes and decode it - fails
LEN=8917
dd if=tmp/decode-test-64kbps.opus of=tmp/truncated-$LEN.opus bs=1 count=$LEN
node dist/test-opus-stream-decoder.js tmp/truncated-$LEN.opus tmp
```

**Comment by AnthumChris** [comment]
Yes, `OpusStreamDecoder`'s underlying `libopusfile` library requires the beginning of the file to discover and instantiate an `OggOpusFile` with [`og_open_*`](https://opus-codec.org/docs/opusfile_api-0.7/group__stream__open__close.html) functions.  Specifically, the header pages and the first audio data page seem to be required, based on my tests below. Data page sizes vary based on the encoding bitrate (higher bitrates, larger pages).  Searching my test file with `xxd` for "OggS" page boundaries, a minimum of 8,918 bytes is required for that file:

```bash
# prep the test files
make test-wasm-module

# inspect Opus file and see the end of first data page at 0x00022d6 (8918 bytes)
xxd -l 8950 -g 8 -c 14 tmp/decode-test-64kbps.opus |grep "OggS"

# create truncated file with 8,918 bytes and decode - succeeds
LEN=8918
dd if=tmp/decode-test-64kbps.opus of=tmp/truncated-$LEN.opus bs=1 count=$LEN
node dist/test-opus-stream-decoder.js tmp/truncated-$LEN.opus tmp

# create truncated file with 8,917 bytes and decode - fails
LEN=8917
dd if=tmp/decode-test-64kbps.opus of=tmp/truncated-$LEN.opus bs=1 count=$LEN
node dist/test-opus-stream-decoder.js tmp/truncated-$LEN.opus tmp
```

**Comment by AnthumChris** [comment]
To clarify above, `libopusfile` requires the header pages and _any_ complete audio data page.  The test below skips the first page and decodes an Opus file with headers and audio page 2:

```bash
# inspect file to see page boundaries for headers and 1st, 2nd audio pages
# Header pages: 0x0-0x349 (0-841 bytes)
# Audio page 1: 0x349-0x22d6 (841-8918 bytes, 8088 total)
# Audo page 2: 0x22d6-0x4263 (8918-16995 bytes, 8078 total)
xxd -l 17010 -g 8 -c 15 tmp/decode-test-64kbps.opus | grep "OggS"

# create file with headers only
OUTFILE="tmp/truncated-headers-page2.opus"
START=0
LEN=`expr 841 - $START`
dd if=tmp/decode-test-64kbps.opus of="$OUTFILE" bs=1 skip="$START" count=$LEN

# append audio page 2 to that file
START=8918
LEN=`expr 16995 - $START`
dd if=tmp/decode-test-64kbps.opus bs=1 skip="$START" count=$LEN >> "$OUTFILE"

# decode the file with headers and page 2 (page 1 skipped)
node dist/test-opus-stream-decoder.js "$OUTFILE" tmp

```

**Comment by zandaqo** [comment]
@AnthumChris To sum up, for streaming we need the header pages and file's total length and duration. Can `libopusfile` isolate those header pages given a chunk of data? I can imagine streaming file by doing two range requests, the first one to get the header pages and the second one to get the audio pages needed. 
Better yet, we can store the header pages alongside other metadata beforehand if there is a sure way of getting them from a file.

**Comment by AnthumChris** [comment]
[`libopusfile.op_test()`](https://mf4.xiph.org/jenkins/view/opus/job/opusfile-unix/ws/doc/html/group__stream__open__close.html#ga54158ed1570a0062cb674115cd3699ab) could test 0-57 range chunk as a valid Opus file, but I didn't see functions in `libopusfile` would actually return the bytes of the parsed headers.  `libogg` may do that or a manual byte scan for page boundaries could suffice.

**Comment by AnthumChris** [comment]
I decided against dynamically allocating a buffer for the entire Opus file size.  It's wasteful and doesn't scale for large Opus files (e.g. a 3-hour SoundCloud DJ mix hundreds of megabytes in size).  I originally considered that option because [`op_open_memory()`](https://mf4.xiph.org/jenkins/view/opus/job/opusfile-unix/ws/doc/html/group__stream__open__close.html#gaffc5769a1e5977f186f77a1fb08cb248) offers an easier development solution.

My current idea for the high-level design is to instantiate a `SeekableDecoder` with a [`Request`](https://developer.mozilla.org/en-US/docs/Web/API/Request) object and specify the time where decoding would begin.  `onDecode()` would receive an additional value `seekStart` to distinguish decoded audio from other callback sequences:

```js
const decoder = new SeekableDecoder(new Request('https://...music.opus'), {onDecode});
decoder.decodeAt(30*1000); // queue to 30 seconds and begin decoding

function onDecode({seekStart, ...}) {}
```

This design removes developers from dealing with the internals of HTTP range requests and parsing, and `decodeAt()` could be called multiple times arbitrarily whenever seeking is needed.  Future performance enhancements could internally cache retrieved HTTP bytes to reduce and optimize HTTP fetches.

**Comment by AnthumChris** [comment]
This feature is currently being developed in branch [http-opus-seek](https://github.com/AnthumChris/opus-stream-decoder/tree/http-opus-seek) and [test-seekable-opus-stream-decoder.html](https://github.com/AnthumChris/opus-stream-decoder/blob/http-opus-seek/dist/test-seekable-opus-stream-decoder.html) is a successful test file.

The duration seeking (finding last `OggS` page byte sequence from end of file) is the most inefficient operation, especially for larger files (I tested with an unrealistic 512 kbit/s file).  Seems best if the duration value were provided during init to avoid duration seeking.  A server-side process that reads the file backwards and sends the value in a header could be cool.  Even cooler if an Nginx module existed to calculate things on the fly and return needed values in HTTP response headers.

**Comment by brion** [comment]
Check out the older oggz-tools package; specifically oggz-chop and its friends which should include support for producing X-Content-Duration headers and doing server-side seek on static files when given a time offset.

Probably not modern, so you may wish to replace them with something that fits into nginx more nicely, but should give you a head start!

**Comment by brion** [comment]
(I'm also not 100% sure if those tools support Opus, but adapting should not be hard if not.)

**Comment by turbo** [comment]
@AnthumChris I want to implement this on a site that streams opus files. All the files have the same sample rate and frame size, and for each file, I have the duration info in the form of time (`4 mins 46.82 secs`) and samples (`13767583`).

Would that simplify the expensive duration seeking you described? If so, how?
